cirq/protocols/json_serialization.py

```diff
@@ -429,6 +446,133 @@ def _cirq_object_hook(d, resolvers: Sequence[JsonResolver]):
     return cls(**d)
 
 
+class SerializableByKey(SupportsJSON):
+    """Protocol for objects that can be serialized to a key + context."""
+
+    @doc_private
+    def _serialization_key_(self) -> str:
+        """Returns a unique string identifier for this object.
+
+        This should only return the same value for two objects if they are
+        equal; otherwise, an error will occur if both are serialized into the
+        same JSON string.
+        """
+
+
+class _SerializedKey(SupportsJSON):
+    """Internal object for holding a SerializableByKey key.
+
+    This is a private type used in contextual serialization. Its deserialization
+    is context-dependent, and is not expected to match the original; in other
+    words, `cls._from_json_dict_(obj._json_dict_())` does not return
+    the original `obj` for this type.
+    """
+
+    def __init__(self, obj: SerializableByKey):
+        self.key = obj._serialization_key_()
+
+    def _json_dict_(self):
+        return obj_to_dict_helper(self, ['key'])
+
+    @classmethod
+    def _from_json_dict_(cls, **kwargs):
+        raise TypeError(f'Internal error: {cls} should never deserialize with _from_json_dict_.')
+
+    @classmethod
+    def read_from_context(cls, context_map, key, **kwargs):
+        return context_map[key]
+
+
+class _SerializedContext(SupportsJSON):
+    """Internal object for a single SerializableByKey key-to-object mapping.
+
+    This is a private type used in contextual serialization. Its deserialization
+    is context-dependent, and is not expected to match the original; in other
+    words, `cls._from_json_dict_(obj._json_dict_())` does not return
+    the original `obj` for this type.
+    """
+
+    def __init__(self, obj: SerializableByKey):
+        self.key = obj._serialization_key_()
+        self.obj = obj
+
+    def _json_dict_(self):
+        return obj_to_dict_helper(self, ['key', 'obj'])
+
+    @classmethod
+    def _from_json_dict_(cls, **kwargs):
+        raise TypeError(f'Internal error: {cls} should never deserialize with _from_json_dict_.')
+
+    @classmethod
+    def update_context(cls, context_map, key, obj, **kwargs):
+        context_map.update({key: obj})
+
+
+class _ContextualSerialization(SupportsJSON):
+    """Internal object for serializing an object with its context.
+
+    This is a private type used in contextual serialization. Its deserialization
+    is context-dependent, and is not expected to match the original; in other
+    words, `cls._from_json_dict_(obj._json_dict_())` does not return
+    the original `obj` for this type.
+    """
+
+    def __init__(self, obj: Any):
+        self.context_list = []
+        context_keys = set()
+        for sbk in get_serializable_by_keys(obj):
+            new_sc = _SerializedContext(sbk)
+            if new_sc.key not in context_keys:
+                self.context_list.append(new_sc)
+                context_keys.add(new_sc.key)
+        self.context_list += [obj]
```

Comment:
Some musings on the naming: I find using the last slot of the context_list confusing a bit for the final object. I looked at a printed json and my mental model after all these names was "context + object" - but then we use _context_list_ that _contains_ the object. So I had to look again, and "aha" I remembered then...

Another mental model could be the "Object Definition DAG" which points towards an ordered list of dependency graph one could be to call it `object_dag` or `objects`. 

Naming it `context` makes it something "meta" instead of the actual objects themselves.

Reply:
I'll change this name to be more accurate.

For clarity: the reason that this _must_ be a list (rather than separate `context` and `object` fields) is to ensure that `read_json` parses context information before the final object. Separate fields could be arbitrarily reordered by JSON; thus the list-plus-object format.

Reply:
Chose `object_dag` and added a clarifying comment.